### ISSUE TITLE
Unneed global variables for StatementInput blocks, procedures blocks, ... ?

### ISSUE THREAD

**Issue by MarkusBordihn** [google/blockly#277]
Simple Example:
https://blockly-demo.appspot.com/static/demos/code/index.html#7z4cm4

This example generates the following JavaScript code:
```javascript
var test;

// Describe this function...
function doSomething(test) {
  if (test <= 10) {
  } else if (test >= 10) {
  }
}
´´´

As you could see the ´´´var test;´´´ is not really needed in this case, because it will be provided by the function itself.
I understand it could be difficult to covering all use cases, but it would be great if you could turn off this auto generated by some kind of xml attribute like ´´´scope="local"´´´ e.g.:
```xml
                  <block type="variables_get">
                      <field name="VAR" scope="local">distance</field>
                  </block>
´´´

**Comment by carlosperate** [comment]
Just wanted to quickly link [this forum discussion](https://groups.google.com/forum/#!topic/blockly/rGMz3NnFVRI) as a background for the current scope situation.

**Comment by MarkusBordihn** [comment]
Thanks for the context. In this case I don't want to define which scope would work the best for Blockly.
I only want to have the possibility to use variables without creating global variables for each possible parameter of an function or event.

The xml attribute should only make sure that no "var test;" is spammed for such context.
It should not provide additional type checking or an additional logic for variables.

But regarding the discussion I think with ES5 and "let" it should be more an mixed mode which should be supported by Blockly for the local and global scope.

**Comment by RoboErikG** [comment]
That might be a side effect of fixing #265 . Something seems wrong, though, as those aren't field_variable entries on the function...

**Comment by RoboErikG** [comment]
Oh! I spoke too soon. That's because the variable needs getters and setters to be accessed and supporting local variables is fairly complicated. For instance, additional UX is needed for local variables to tell users why they can only use them within scope, and detecting that is another chunk of logic.

The App Inventor project has an implementation of local scoping that you could look at if you're interested.

**Comment by carlosperate** [comment]
Indeed, you'd need to add checks for ensuring the variable is only used within the defining function, and you would need to add extra rules to the variable field dropdown to include the relevant available variables depending on scope.

**Comment by aminmarashi** [comment]
In my opinion, it's not a good idea to provide setters for function args. It's best if you could treat the function args differently than variables. They can have different blocks for themselves. 
Reasons why it's not a good idea:
1. Defining a variable with the same name as a function argument is possible and very confusing for the user. [take a look](https://blockly-demo.appspot.com/static/demos/code/index.html#dcw9s9)
2. Setting the variable inside the function doesn't have the intended effect. Users may get confused and do sth like [this](https://blockly-demo.appspot.com/static/demos/code/index.html#kqa5a2)
I think the best way is to control the arg variables inside the var editor menu of the function provide getters only, not setters.

**Comment by TimNZ** [comment]
Slightly off topic but related - support to prevent declaration of a variable?

I want to take JS generated code as is and wrap it in a function that takes args I will provide values for. e.g. 
"function(test) { <generated code that does something with 'test'> }

Looks like variables just need to have a flag that can be checked in generator init() function to exclude a variable from being declared.
i.e. Blockly.Javascript.init() - modify getAllVariables() loop

**Comment by RoboErikG** [comment]
Hi GreenCheek, that kind of question would be better asked on [the forum](https://groups.google.com/forum/#!forum/blockly) instead of an issue.

**Comment by GreenCheek** [comment]
RoboErikG thank you, I'll try my luck there.

**Comment by MNSDivya** [comment]
var test; is still generated. 
![image](https://user-images.githubusercontent.com/56336353/81164153-844d6400-8fad-11ea-90cb-ad282da02dc3.png)

Issue reported still persists
